### PR TITLE
Fix unicode redirects on Python 3.

### LINKED ISSUE #2653: .htaccesss redirect to non ASCII folder does not work
Hello,

I have the following setup on a shared hoster:
- Apache 2.2.15
- A Japanese language .みんな (.minna; xn--q9jyb4c) IDN domain.
- A blog which is in the subfolder ブログ (blog)
- A redirect in the .htaccess file like this: `Redirect /index.html /ブログ/`

So I usually open the domain http://test.みんな and the server redirects to http://test.みんな/ブログ. This works fine in Firefox etc.

With requests, I get the following error (Python 3.4 with Requests 2.7.0 on a Japanese Ubuntu 15.04):

```
'<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">\n<html><head>\n<title>404 Not Found</title>\n</head><body>\n<h1>Not Found</h1>\n<p>The requested URL /Ã£Â\x83Â\x96Ã£Â\x83Â\xadÃ£Â\x82Â°/ was not found on this server.</p>\n<hr>\n<address>Apache/2.2.15 (CentOS) Server at test.xn--q9jyb4c Port 80</address>\n</body></html>\n'
```

So I guess the request lib gets a redirect from a server with Japanese characters, but then fails to convert the characters correctly. If I do `requests.get(http://test.みんな/ブログ)` directly it works, only the redirect does not.

### PR DESCRIPTION
**Pull request by Lukasa**
Resolves #2653.

Follows on from #2655: please read the discussion there for more information.

### CODE REVIEW

**Comment by kennethreitz** [comment]
:+1:

**Comment by kennethreitz** [comment]
there's a lot of pending merges — trying hard not to merge them :)

**Comment by Lukasa** [comment]
@kennethreitz You're doing an admirable job. ;)

**Comment by sigmavirus24** [comment]
Soon™

**Comment by sigmavirus24** [comment]
Oh right, this is for 3.0.0... did we have a test for this? If not I'm :+1:

**Comment by Lukasa** [comment]
I don't think we do, though I can see if I can create one.

**Comment by sigmavirus24** [comment]
Tests are nice :-P

**Comment by Lukasa** [comment]
Ok, got some extra tests for `unquote_unreserved`, which is a start.

**Comment by Lukasa** [comment]
And a failing test for the original issue, which now passes. It's only valid on Python 3, so I added a skip for Python 2.

Unfortunately, this now requires a more up-to-date py.test version.

**Comment by Lukasa** [comment]
Hurrah! All tests green. Over to you @sigmavirus24.

File: requests/utils.py [hunk 1]
```diff
@@ -396,7 +397,28 @@ def unquote_unreserved(uri):
...
+        elif is_py3 and is_bytes:
+            return c.encode('ascii')
+        else:
+            return c
+
+    # Handle both bytestrings and unicode strings.
+    if isinstance(uri, bytes):
+        splitchar = b'%'
+        base = b''
+        convert = functools.partial(_convert, True)
```
Comment (sigmavirus24) [inline requests/utils.py]: We could do this another way potentially (although I'm not certain it's necessarily better):

``` py
is_bytes = isinstance(uri, bytes)
splitchar = u'%'
base = u''
if is_bytes:
    splitchar = splitchar.encode('ascii')
    base = base.encode('ascii')

# And then down below

if c in UNRESERVED_SET:
    parts[i] = convert(is_bytes, c) + parts[i][2:]
```

**Comment by sigmavirus24** [comment]
One tiny comment. I'd like your thoughts on it (not necessarily action unless you agree). Otherwise :+1:

File: requests/utils.py (hunk 1 above)
Comment (Lukasa) [inline requests/utils.py]: I think this is better, I'll do this shortly.

Comment (Lukasa) [inline requests/utils.py]: Done.

File: requests/utils.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import codecs
 import collections
 import io
+import functools
```
Comment (sigmavirus24) [inline requests/utils.py]: Do we still need functools in here?

Comment (kennethreitz) [inline requests/utils.py]: ping — six months of no progress

**Comment by kennethreitz** [comment]
Rebase required.

**Comment by Lukasa** [comment]
Rebased.

**Pull request merged**